src/openrct2/actions/LandBuyRightsAction.hpp

```diff
@@ -0,0 +1,185 @@
+    {
+    }
+    LandBuyRightsAction(MapRange range, LandBuyRightSetting setting)
+        : _range(range)
+        , _setting(static_cast<uint8_t>(setting))
+    {
+    }
+
+    LandBuyRightsAction(CoordsXY coord, LandBuyRightSetting setting)
+        : _range(coord.x, coord.y, coord.x, coord.y)
+        , _setting(static_cast<uint8_t>(setting))
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags();
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << DS_TAG(_range) << DS_TAG(_setting);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        return QueryExecute(false);
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        return QueryExecute(true);
+    }
+
+private:
+    GameActionResult::Ptr QueryExecute(bool isExecuting) const
+    {
+        auto res = MakeResult();
+
+        MapRange normRange = _range.Normalise();
+        // Keep big coordinates within map boundaries
+        auto aX = std::max<decltype(normRange.GetLeft())>(64, normRange.GetLeft());
+        auto bX = std::min<decltype(normRange.GetRight())>(gMapSizeMaxXY - 32, normRange.GetRight());
+        auto aY = std::max<decltype(normRange.GetTop())>(64, normRange.GetTop());
+        auto bY = std::min<decltype(normRange.GetBottom())>(gMapSizeMaxXY - 32, normRange.GetBottom());
+
+        MapRange validRange = MapRange{ aX, aY, bX, bY };
+
+        CoordsXYZ centre{ (validRange.GetLeft() + validRange.GetRight()) / 2 + 16,
+                          (validRange.GetTop() + validRange.GetBottom()) / 2 + 16, 0 };
+        centre.z = tile_element_height(centre.x, centre.y) & 0xFFFF;
+
+        res->Position = centre;
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_LAND_PURCHASE;
+
+        // Game command modified to accept selection size
+        for (auto y = validRange.GetTop(); y <= validRange.GetBottom(); y += 32)
+        {
+            for (auto x = validRange.GetLeft(); x <= validRange.GetRight(); x += 32)
+            {
+                auto result = map_buy_land_rights_for_tile({ x, y }, isExecuting);
+                if (result->Error == GA_ERROR::OK)
+                {
+                    res->Cost += result->Cost;
+                }
+            }
+        }
+        if (isExecuting)
+        {
+            map_count_remaining_land_rights();
+        }
+        return res;
+    }
+
+    GameActionResult::Ptr map_buy_land_rights_for_tile(const CoordsXY loc, bool isExecuting) const
+    {
+        if (_setting >= static_cast<uint8_t>(LandBuyRightSetting::Count))
+        {
+            log_warning("Tried calling buy land rights with an incorrect setting. setting = %u", _setting);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, _ErrorTitles[0], STR_NONE);
+        }
+
+        SurfaceElement* surfaceElement = map_get_surface_element_at(loc)->AsSurface();
```

Comment:
If map_get_surface_element_at returns nullptr this will try to dereference null when calling AsSurface()